assembly/helpers/liquidityManager.ts

```diff
@@ -0,0 +1,276 @@
+import { fromBytes, toBytes } from '@massalabs/as-types';
+import { Address, Storage } from '../std';
+
+// to move in it's dedicated file
+export class StoragePrefixManager {
+  private prefix: u8 = 0;
+  newPrefix(): u8 {
+    return this.prefix++;
+  }
+}
+
+// Key used to store the total supply in storage.
+const totalSupplyKey: u8 = 0x00;
+
+/**
+ * Manages liquidity for tokens, with generic flexibility.
+ *
+ * @remarks
+ * It does not implement overflow and underflow checks for operations.
+ *
+ * @typeParam T - The type used for liquidity amount.
+ *
+ * @privateRemarks
+ * Balance and allowance are stored separately in storage using distinct prefixes.
+ * Total supply is stored under a fixed key.
+ */
+export class LiquidityManager<T> {
+  private balancePrefix: u8;
+  private allowancePrefix: u8;
+
+  constructor(storagePrefixManager: StoragePrefixManager) {
+    this.balancePrefix = storagePrefixManager.newPrefix();
+    this.allowancePrefix = storagePrefixManager.newPrefix();
+  }
+
+  @inline
+  private _getOrNull(key: StaticArray<u8>): T {
+    return Storage.has(key) ? <T>fromBytes<T>(Storage.get(key)) : <T>0;
+  }
+
+  @inline
+  private _free(key: StaticArray<u8>): void {
+    if (Storage.has(key)) {
+      assert(
+        fromBytes<T>(Storage.get(key)) == <T>0,
+        `Storage value is not null ${fromBytes<T>(Storage.get(key))}`,
+      );
+      Storage.del(key);
+    }
+  }
+
+  @inline
+  private _getTotalSupplyKey(): StaticArray<u8> {
+    return [this.balancePrefix, totalSupplyKey];
+  }
+
+  /**
+   * Fetches the total supply of the token.
+   *
+   * @remarks
+   * By default, the total supply is 0.
+   *
+   * @returns The total token supply.
+   */
+  public getTotalSupply(): T {
+    return this._getOrNull(this._getTotalSupplyKey());
+  }
+
+  @inline
+  private _getBalanceStorageKey(user: Address): StaticArray<u8> {
+    const key = new StaticArray<u8>(1);
+    key[0] = this.balancePrefix;
+    return key.concat(user.serialize());
+  }
+
+  private _updateAmount(key: StaticArray<u8>, amount: T, increase: bool): void {
+    var newAmount = this._getOrNull(key);
+    if (increase) {
+      // @ts-ignore arithmetic operations on generic types
+      newAmount += amount;
+    } else {
+      // @ts-ignore arithmetic operations on generic types
+      newAmount -= amount;
+    }
+    Storage.set(key, toBytes(newAmount));
+  }
+
+  /**
+   * Retrieves the balance for a specified user address.
+   *
+   * @param user - The user's address.
+   *
+   * @returns User's balance, or 0 if no balance is found.
+   */
+  public getBalance(user: Address): T {
```

Comment:
People are used to `balanceOf` bu it doesn't make such a difference

Reply:
Understood, but then, following this logic, shouldn't we rename `mint` and `burn` with `issue` and `destroy` ?

Reply:
Because I like mint and burn. I find it explicit.

Reply:
But then you are potentially mixing up smart contract exported functions and object methods